File: typescript/tests/test_http_signer.ts
import { describe, it, beforeEach, afterEach } from "node:test";
import assert from "node:assert/strict";
import { createServer, type Server, type IncomingMessage, type ServerResponse } from "node:http";
import { inspect } from "node:util";

import { HttpSigner } from "../src/http-signer.js";

// ─── Mock Signer Server ──────────────────────────────────────────────────────

const MOCK_ADDRESS = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
const MOCK_SIGNATURE = "0x" + "ab".repeat(32) + "cd".repeat(32) + "1b";
const VALID_TOKEN = "rmit_sk_" + "a1".repeat(32);

interface MockServerOptions {
  /** Override the response for specific paths. */
  overrides?: Record<string, { status: number; body: unknown }>;
}

function createMockServer(options?: MockServerOptions): { server: Server; port: number; url: string } {
  const overrides = options?.overrides ?? {};
  const server = createServer((req: IncomingMessage, res: ServerResponse) => {
    const path = req.url ?? "/";

    // Check auth on all endpoints except /health
    if (path !== "/health") {
      const auth = req.headers.authorization;
      if (!auth || !auth.startsWith("Bearer ")) {
        res.writeHead(401, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ error: "unauthorized" }));
        return;
      }
      const token = auth.slice(7);
      if (token !== VALID_TOKEN) {
        res.writeHead(401, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ error: "unauthorized" }));
        return;
      }
    }

    // Check overrides first
    if (overrides[path]) {
      const { status, body } = overrides[path]!;
      res.writeHead(status, { "Content-Type": "application/json" });
      res.end(JSON.stringify(body));
      return;
    }

    if (path === "/address" && req.method === "GET") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ address: MOCK_ADDRESS }));
    } else if (path === "/sign/typed-data" && req.method === "POST") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ signature: MOCK_SIGNATURE }));
    } else if (path === "/health" && req.method === "GET") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ ok: true, version: "test", wallet: "test" }));
    } else {
      res.writeHead(404, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ error: "not_found" }));
    }
  });

  // Listen on random port
  server.listen(0);
  const addr = server.address();
  const port = typeof addr === "object" && addr !== null ? addr.port : 0;
  return { server, port, url: `http://127.0.0.1:${port}` };
}

// ─── Tests ───────────────────────────────────────────────────────────────────

describe("HttpSigner", () => {
  let server: Server;
  let url: string;

  beforeEach(() => {
    const mock = createMockServer();
    server = mock.server;
    url = mock.url;
  });

  afterEach(() => {
    server.close();
  });

  // ── Construction ──────────────────────────────────────────────────────

  it("create() fetches and caches address", async () => {
    const signer = await HttpSigner.create({ url, token: VALID_TOKEN });
    assert.equal(signer.getAddress(), MOCK_ADDRESS);
  });

  it("getAddress() throws if not initialized", () => {
    // Can't call constructor directly (private), but we can test the error path
    // by checking that create() is required
    assert.ok(HttpSigner.create, "create static method must exist");
  });

  // ── signTypedData ─────────────────────────────────────────────────────

  it("signTypedData() returns signature from server", async () => {
    const signer = await HttpSigner.create({ url, token: VALID_TOKEN });
    const sig = await signer.signTypedData(
      { name: "Test", version: "1" },
      { Test: [{ name: "value", type: "uint256" }] },
      { value: 42 },
    );
    assert.equal(sig, MOCK_SIGNATURE);
  });

  it("signTypedData() handles BigInt values", async () => {
    const signer = await HttpSigner.create({ url, token: VALID_TOKEN });
    // BigInt should be serialized as string without throwing
    const sig = await signer.signTypedData(
      { name: "Test", version: "1" },
      { Test: [{ name: "value", type: "uint256" }] },
      { value: BigInt("1000000000000000000") },
    );
    assert.equal(sig, MOCK_SIGNATURE);
  });

  // ── Auth errors ───────────────────────────────────────────────────────

  it("create() fails with bad token", async () => {
    await assert.rejects(
      () => HttpSigner.create({ url, token: "bad_token_abc" }),
      (err: Error) => {
        assert.ok(err.message.includes("401"), `expected 401 mention, got: ${err.message}`);
        return true;
      },
    );
  });

  it("signTypedData() throws on 401", async () => {
    // Create with valid token, then manually test with a server that rejects
    const mock2 = createMockServer({
      overrides: {
        "/address": { status: 200, body: { address: MOCK_ADDRESS } },
        "/sign/typed-data": { status: 401, body: { error: "unauthorized" } },
      },
    });
    const signer = await HttpSigner.create({ url: mock2.url, token: VALID_TOKEN });

    await assert.rejects(
      () => signer.signTypedData({ name: "T", version: "1" }, {}, {}),
      (err: Error) => {
        assert.ok(err.message.includes("unauthorized"), `expected unauthorized, got: ${err.message}`);
        return true;
      },
    );
    mock2.server.close();
  });

  // ── Policy denial ─────────────────────────────────────────────────────

  it("signTypedData() throws on 403 with reason", async () => {
    const mock2 = createMockServer({
      overrides: {
        "/address": { status: 200, body: { address: MOCK_ADDRESS } },
        "/sign/typed-data": { status: 403, body: { error: "policy_denied", reason: "chain not allowed" } },
      },
    });
    const signer = await HttpSigner.create({ url: mock2.url, token: VALID_TOKEN });

    await assert.rejects(
      () => signer.signTypedData({ name: "T", version: "1" }, {}, {}),
      (err: Error) => {
        assert.ok(err.message.includes("policy denied"), `expected policy denied, got: ${err.message}`);
        assert.ok(err.message.includes("chain not allowed"), `expected reason, got: ${err.message}`);
        return true;
      },
    );
    mock2.server.close();
  });

  // ── Server error ──────────────────────────────────────────────────────

  it("signTypedData() throws on 500", async () => {
    const mock2 = createMockServer({
      overrides: {
        "/address": { status: 200, body: { address: MOCK_ADDRESS } },
        "/sign/typed-data": { status: 500, body: { error: "internal_error" } },
      },
    });
    const signer = await HttpSigner.create({ url: mock2.url, token: VALID_TOKEN });

    await assert.rejects(
      () => signer.signTypedData({ name: "T", version: "1" }, {}, {}),
      (err: Error) => {
        assert.ok(err.message.includes("500"), `expected 500, got: ${err.message}`);
        return true;
      },
    );
    mock2.server.close();
  });

  // ── Server unreachable ────────────────────────────────────────────────

  it("create() throws when server unreachable", async () => {
    await assert.rejects(
      () => HttpSigner.create({ url: "http://127.0.0.1:1", token: VALID_TOKEN }),
      (err: Error) => {
        assert.ok(err.message.includes("cannot reach"), `expected cannot reach, got: ${err.message}`);
        return true;
      },
    );
  });

  // ── No credential leakage ─────────────────────────────────────────────

  it("toJSON() does not leak token", async () => {
    const signer = await HttpSigner.create({ url, token: VALID_TOKEN });
    const json = signer.toJSON();
    assert.equal(json.address, MOCK_ADDRESS);
    assert.ok(!JSON.stringify(json).includes(VALID_TOKEN), "token must not appear in JSON");
  });

  it("inspect() does not leak token", async () => {
    const signer = await HttpSigner.create({ url, token: VALID_TOKEN });
    const str = inspect(signer);
    assert.ok(!str.includes(VALID_TOKEN), "token must not appear in inspect output");
    assert.ok(str.includes(MOCK_ADDRESS), "address should appear in inspect");
  });

  // ── Malformed response ────────────────────────────────────────────────

  it("signTypedData() throws on malformed response", async () => {
    const mock2 = createMockServer({
      overrides: {
        "/address": { status: 200, body: { address: MOCK_ADDRESS } },
        "/sign/typed-data": { status: 200, body: { notSignature: true } },
      },
    });
    const signer = await HttpSigner.create({ url: mock2.url, token: VALID_TOKEN });

    await assert.rejects(
      () => signer.signTypedData({ name: "T", version: "1" }, {}, {}),
      (err: Error) => {
        assert.ok(err.message.includes("no signature"), `expected no signature, got: ${err.message}`);
        return true;
      },
    );
    mock2.server.close();
  });
});

File: python/tests/test_http_signer.py
"""Tests for the HTTP signer adapter."""

from __future__ import annotations

import json
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import ClassVar

import pytest

from remitmd.http_signer import HttpSigner

# ── Mock signer server ──────────────────────────────────────────────────────

MOCK_ADDRESS = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"
MOCK_SIGNATURE = "0x" + "ab" * 32 + "cd" * 32 + "1b"
VALID_TOKEN = "rmit_sk_" + "a1" * 32


class MockSignerHandler(BaseHTTPRequestHandler):
    """HTTP handler that mimics the local signer server."""

    # Class-level overrides for specific paths
    overrides: ClassVar[dict[str, tuple[int, dict[str, object]]]] = {}

    def log_message(self, format: str, *args: object) -> None:
        pass  # Suppress logs

    def _check_auth(self) -> bool:
        auth = self.headers.get("Authorization", "")
        if not auth.startswith("Bearer "):
            self._respond(401, {"error": "unauthorized"})
            return False
        token = auth[7:]
        if token != VALID_TOKEN:
            self._respond(401, {"error": "unauthorized"})
            return False
        return True

    def _respond(self, status: int, body: dict[str, object]) -> None:
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(body).encode())

    def do_GET(self) -> None:
        path = self.path

        if path in self.overrides:
            status, body = self.overrides[path]
            self._respond(status, body)
            return

        if path == "/health":
            self._respond(200, {"ok": True})
            return

        if not self._check_auth():
            return

        if path == "/address":
            self._respond(200, {"address": MOCK_ADDRESS})
        else:
            self._respond(404, {"error": "not_found"})

    def do_POST(self) -> None:
        path = self.path

        if path in self.overrides:
            status, body = self.overrides[path]
            self._respond(status, body)
            return

        if not self._check_auth():
            return

        if path == "/sign/typed-data":
            self._respond(200, {"signature": MOCK_SIGNATURE})
        else:
            self._respond(404, {"error": "not_found"})


def start_mock_server(
    overrides: dict[str, tuple[int, dict[str, object]]] | None = None,
) -> tuple[HTTPServer, str]:
    """Start a mock signer server on a random port."""

    class Handler(MockSignerHandler):
        pass

    Handler.overrides = overrides or {}
    server = HTTPServer(("127.0.0.1", 0), Handler)
    port = server.server_address[1]
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server, f"http://127.0.0.1:{port}"


# ── Fixtures ─────────────────────────────────────────────────────────────────


@pytest.fixture()
def mock_server():
    server, url = start_mock_server()
    yield url
    server.shutdown()


@pytest.fixture()
def policy_denied_server():
    server, url = start_mock_server(
        overrides={
            "/sign/typed-data": (403, {"error": "policy_denied", "reason": "chain not allowed"}),
        }
    )
    yield url
    server.shutdown()


@pytest.fixture()
def error_server():
    server, url = start_mock_server(
        overrides={
            "/sign/typed-data": (500, {"error": "internal_error"}),
        }
    )
    yield url
    server.shutdown()


@pytest.fixture()
def malformed_server():
    server, url = start_mock_server(
        overrides={
            "/sign/typed-data": (200, {"not_signature": True}),
        }
    )
    yield url
    server.shutdown()


# ── Tests ────────────────────────────────────────────────────────────────────


class TestHttpSignerCreate:
    @pytest.mark.anyio()
    async def test_create_fetches_address(self, mock_server: str) -> None:
        signer = await HttpSigner.create(mock_server, VALID_TOKEN)
        assert signer.get_address() == MOCK_ADDRESS

    @pytest.mark.anyio()
    async def test_create_bad_token(self, mock_server: str) -> None:
        with pytest.raises(PermissionError, match="unauthorized"):
            await HttpSigner.create(mock_server, "bad_token")

    @pytest.mark.anyio()
    async def test_create_unreachable(self) -> None:
        with pytest.raises(ConnectionError, match="cannot reach"):
            await HttpSigner.create("http://127.0.0.1:1", VALID_TOKEN)


class TestHttpSignerSign:
    @pytest.mark.anyio()
    async def test_sign_returns_signature(self, mock_server: str) -> None:
        signer = await HttpSigner.create(mock_server, VALID_TOKEN)
        sig = await signer.sign_typed_data(
            {"name": "Test", "version": "1"},
            {"Test": [{"name": "value", "type": "uint256"}]},
            {"value": 42},
        )
        assert sig == MOCK_SIGNATURE

    @pytest.mark.anyio()
    async def test_sign_policy_denied(self, policy_denied_server: str) -> None:
        signer = await HttpSigner.create(policy_denied_server, VALID_TOKEN)
        with pytest.raises(PermissionError, match="policy denied"):
            await signer.sign_typed_data({}, {}, {})

    @pytest.mark.anyio()
    async def test_sign_server_error(self, error_server: str) -> None:
        signer = await HttpSigner.create(error_server, VALID_TOKEN)
        with pytest.raises(RuntimeError, match="500"):
            await signer.sign_typed_data({}, {}, {})

    @pytest.mark.anyio()
    async def test_sign_malformed_response(self, malformed_server: str) -> None:
        signer = await HttpSigner.create(malformed_server, VALID_TOKEN)
        with pytest.raises(RuntimeError, match="no signature"):
            await signer.sign_typed_data({}, {}, {})


class TestHttpSignerNoLeakage:
    @pytest.mark.anyio()
    async def test_repr_no_token(self, mock_server: str) -> None:
        signer = await HttpSigner.create(mock_server, VALID_TOKEN)
        r = repr(signer)
        assert VALID_TOKEN not in r
        assert MOCK_ADDRESS in r

    @pytest.mark.anyio()
    async def test_str_no_token(self, mock_server: str) -> None:
        signer = await HttpSigner.create(mock_server, VALID_TOKEN)
        s = str(signer)
        assert VALID_TOKEN not in s


class TestHttpSignerNotInitialized:
    def test_get_address_before_create(self) -> None:
        signer = HttpSigner("http://localhost:7402", "token")
        with pytest.raises(RuntimeError, match="not initialized"):
            signer.get_address()
